ampd-handlers/src/bin/evm/handler.rs

```diff
@@ -0,0 +1,214 @@
+use cosmrs::tx::Msg;
+use cosmrs::{AccountId, Any};
+use error_stack::ResultExt;
+use ethers_core::types::{TransactionReceipt, U64};
+use events::{try_from, AbciEventTypeFilter};
+use futures::future::join_all;
+use serde::Deserialize;
+use tracing::{info, info_span};
+use typed_builder::TypedBuilder;
+use valuable::Valuable;
+use voting_verifier::msg::ExecuteMsg;
+
+use crate::Error;
+
+type Result<T> = error_stack::Result<T, Error>;
+
+#[derive(Clone, Debug, Deserialize)]
+#[try_from("wasm-messages_poll_started")]
+pub struct PollStartedEvent {
+    poll_id: PollId,
+    source_chain: ChainName,
+    source_gateway_address: EVMAddress,
+    confirmation_height: u64,
+    expires_at: u64,
+    messages: Vec<Message>,
+    participants: Vec<AccountId>,
+}
+
+#[derive(Debug, TypedBuilder)]
+pub struct Handler<C>
+where
+    C: EthereumClient,
+{
+    verifier: AccountId,
+    voting_verifier_contract: AccountId,
+    chain: ChainName,
+    finalizer_type: Finalization,
+    rpc_client: C,
+    monitoring_client: monitoring::Client,
+}
+
+impl<C> Handler<C>
+where
+    C: EthereumClient + Send + Sync,
+{
+    async fn finalized_tx_receipts<T>(
+        &self,
+        tx_hashes: T,
+        confirmation_height: u64,
+    ) -> Result<HashMap<Hash, TransactionReceipt>>
+    where
+        T: IntoIterator<Item = Hash>,
+    {
+        let latest_finalized_block_height =
+            finalizer::pick(&self.finalizer_type, &self.rpc_client, confirmation_height)
+                .latest_finalized_block_height()
+                .await
+                .change_context(Error::Finalizer)?;
+
+        Ok(join_all(
+            tx_hashes
+                .into_iter()
+                .map(|tx_hash| self.rpc_client.transaction_receipt(tx_hash)),
+        )
+        .await
+        .into_iter()
+        .filter_map(std::result::Result::unwrap_or_default)
+        .filter_map(|tx_receipt| {
+            if tx_receipt
+                .block_number
+                .unwrap_or(U64::MAX)
+                .le(&latest_finalized_block_height)
+            {
+                Some((tx_receipt.transaction_hash, tx_receipt))
+            } else {
+                None
+            }
+        })
+        .collect())
+    }
+
+    fn vote_msg(&self, poll_id: PollId, votes: Vec<Vote>) -> MsgExecuteContract {
+        MsgExecuteContract {
+            sender: self.verifier.clone(),
+            contract: self.voting_verifier_contract.clone(),
+            msg: serde_json::to_vec(&ExecuteMsg::Vote { poll_id, votes })
+                .expect("vote msg should serialize"),
+            funds: vec![],
+        }
+    }
+}
+
+#[async_trait]
+impl<C> EventHandler for Handler<C>
+where
+    C: EthereumClient + Send + Sync,
+{
+    type Err = Error;
+    type Event = PollStartedEvent;
+
+    async fn handle<HC: EventHandlerClient + Send + 'static>(
+        &self,
+        event: PollStartedEvent,
+        client: &mut HC,
+    ) -> Result<Vec<Any>> {
```

Comment:
note to reviewer: removed contract check because the filter is now doing that. This also removes the need of extracting `_contract_address` from `PollStartedEvent` because we don't have the access to the raw Event at this point like with the old handler

Reply:
I believe in the subscribe function there is still a dummy value, `contract: Default::default()`. So filtering by contract is actually not handled yet